### PR TITLE
Likeable refactor

### PR DESCRIPTION
**Pull request by ProCynic**
User owner pulled up into Likeable instead of being in subclasses.  All references changed.
Likes class removed, replaced with ManyToMany association.
Visibility enum added to Likeable, defaults to friends.
When creating anything that inherits from likeable, you can supply a visibility, and then query against it when loading from db.
You can also use the Visible(User user) method of Likeable to check if the Likeable is visible to a given user.

### CODE REVIEW

File: app/controllers/Notes.java [hunk 1]
```diff
@@ -9,24 +9,23 @@
...
         render(user);
     }
-    
+
     public static void saveNote(String title, String content) {
-    	
-    	Note n = new Note(user(), title, content);
-    	n.save();
-    	viewNotes();
+
```
Comment (amshali) [inline app/controllers/Notes.java]: .... indentation....

File: app/controllers/AlbumController.java [hunk 2]
```diff
@@ -130,16 +127,16 @@ public static boolean updateDate(Long Id){
...
    */
   public static List<Album> getAll(){
     return Album.findAll();// no idea how this works???
   }
 
-  
+
   /**getNext
-   * 
+   *
```
Comment (amshali) [inline app/controllers/AlbumController.java:139]: Lots of changes that I don't get.

File: app/controllers/Comments.java [hunk 3]
```diff
@@ -25,46 +37,27 @@ public static void comments(Long userId) {
...
-  public static void postComment(Long statusId, Long userId, String commentContent) {
-    ((Status) Status.findById(statusId)).addComment(Application.user(), commentContent);
-    comments(userId);
+    if (c == null)
+      notFound("That comment does not exist.");
+    c.delete();
+    ok();
   }
 
-  public static void makeNewComment(String commentContent, String statusId, String userId) {
```
Comment (amshali) [inline app/controllers/Comments.java:60]: Where did this go?

File: app/controllers/Events.java [hunk 4]
```diff
@@ -95,51 +95,51 @@ public static void deleteEvent(Long eventId) {
...
-			} else if (curEvent.privilege.equals("inviteOnly")) {
-				event.inviteOnly = true;
-			}
-			event.members = new ArrayList<User>();
-			event.members.add(currentUser);
-			event.save();
-			event(event.id);
-		}
-	}
+                User currentUser = user();
```
Comment (amshali) [inline app/controllers/Events.java]: So, what was wrong with the previous indentation here?

File: app/controllers/Events.java [hunk 5]
```diff
@@ -175,10 +175,10 @@ public static void newEventPost(Long eventId, Long userId, String post_content){
     new Post((Event)Event.findById(eventId), (User)User.findById(userId), post_content).save();
     event(eventId);
   }
-  
+
   public static void leaveEvent(Long eventId, Long userId){
     User guest = User.findById(userId);
-    Event event = Event.findById(eventId);
```
Comment (amshali) [inline app/controllers/Events.java:181]: useless change.

File: app/controllers/Photos.java [hunk 6]
```diff
@@ -248,14 +237,14 @@ public static void setGravatar(String gravatarEmail) throws FileNotFoundExceptio
           //create new blob
           Blob blob = new Blob();
           blob.set(new FileInputStream(gravatar),
-              MimeTypes.getContentType(gravatar.getName()));
+                   MimeTypes.getContentType(gravatar.getName()));
```
Comment (amshali) [inline app/controllers/Photos.java:240]: ?

File: app/controllers/Profiles.java [hunk 7]
```diff
@@ -156,10 +156,11 @@ public static void updateBasic(String religion, String bio, Date birthday, Strin
...
     profile.work = work;
-  //  u.profile.hometown = hometown;
+    //  u.profile.hometown = hometown;
     profile.quotes = quotes;
     profile.save();
     user.save();
     renderTemplate("Application/edit_basic.html", profile);
-  }*/
+    }
+  */
```
Comment (amshali) [inline app/controllers/Profiles.java:165]: They are changing this file and these changes are gonna just confuse them.

File: app/models/Profile.java [hunk 8]
```diff
@@ -49,14 +49,47 @@
 
   public String political; // The user's political view
   public String quotes; // The user's favorite quotes
-  public String relationshipStatus; // The user's relationship
+
+  public enum Relationship {
```
Comment (amshali) [inline app/models/Profile.java:53]: Like this.

File: app/controllers/AlbumController.java (hunk 2 above)
Comment (ProCynic) [inline app/controllers/AlbumController.java:139]: I don't know, looks like I somehow removed a bunch of spaces from the end of some lines?

Comment (jtessler) [inline app/controllers/AlbumController.java:139]: Looks like he ran a script to remove all trailing whitespace. A nice thing to do, but it makes reviewing the pull request harder...

File: app/controllers/Comments.java (hunk 3 above)
Comment (ProCynic) [inline app/controllers/Comments.java:60]: app/controllers/Commentable.java

File: app/controllers/Events.java (hunk 4 above)
Comment (ProCynic) [inline app/controllers/Events.java]: don't know, might be some sort of merge thing.

**Comment by amshali** [comment]
I would say there are good amount of good refactorings here.  But, there are some changes that is going to confuse others at this stage. Because they are trying to get things and if we merge this they have to pull again and merge and all that. Can you break this pull request into a bunch of smaller ones?

File: app/controllers/Photos.java (hunk 6 above)
Comment (ProCynic) [inline app/controllers/Photos.java:240]: That's the proper indentation apparently.  emacs does a good job of setting it unless you have annotations.

File: app/controllers/AlbumController.java (hunk 2 above)
Comment (ProCynic) [inline app/controllers/AlbumController.java:139]: I didn't run a script.  Might be emacs automatic indentation fix.

File: app/controllers/Profiles.java (hunk 7 above)
Comment (ProCynic) [inline app/controllers/Profiles.java:165]: This is merged with the latest version.

File: app/controllers/Notes.java (hunk 1 above)
Comment (ProCynic) [inline app/controllers/Notes.java]: fixed

File: app/controllers/AlbumController.java (hunk 2 above)
Comment (ProCynic) [inline app/controllers/AlbumController.java:139]: also, I run M-x untabify on saving.

File: app/models/Note.java [hunk 9]
```diff
@@ -28,18 +28,18 @@ public Note(User author, String title, String content) {
...
-	  if (this.content.length() < TEASER_LENGTH) {
-		  return this.content;
-	  } else {
-		  return this.content.substring(0, TEASER_LENGTH);
-	  }
+    if (this.content.length() < TEASER_LENGTH) {
+      return this.content;
+    } else {
+      return this.content.substring(0, TEASER_LENGTH);
+    }
```
Comment (ProCynic) [inline app/models/Note.java:37]: I think this might be replacing tabs with spaces.  Not sure.

File: app/models/Post.java [hunk 10]
```diff
@@ -35,24 +33,20 @@
   private static final int TEASER_LENGTH = 150;
 
   public Post(Postable postedObj, User author, String content) {
+    super(author);
     this.postedObj = postedObj;
     this.tags = new TreeSet<Tag>();
     this.mentions = new ArrayList<User>();
-    this.author = author;
-    this.content = parseContent(content);
+    this.content = parseContent(HTML.htmlEscape(content));
```
Comment (ProCynic) [inline app/models/Post.java:40]: Moved html escaping into Post constructor.  Similar change in comment.

File: app/models/Profile.java (hunk 8 above)
Comment (ProCynic) [inline app/models/Profile.java:53]: you'll notice that the constructor just uses fromString, so if the string you pass in isn't one of these you'll get a null I think.

**Comment by amshali** [comment]
Looks good to me. I guess others have to deal with it.

**Comment by ProCynic** [comment]
I really would have liked to have put this through a few days ago but it took me a while to find that bug where I was using the default join table for 2 separate associations between the same two classes.  Also, I ended up having to redo the post and comment js after my back end changes broke it.

**Comment by amshali** [comment]
Great job!

**Pull request merged**